src/librustdoc/clean/types.rs

```diff
@@ -1394,33 +1408,50 @@ impl Type {
     }
 }
 
-impl GetDefId for Type {
-    fn def_id(&self) -> Option<DefId> {
+impl Type {
+    fn inner_def_id(&self, cache: Option<&Cache>) -> Option<DefId> {
+        fn inner<T: GetDefId>(t: &T, cache: Option<&Cache>) -> Option<DefId> {
+            match cache {
+                Some(c) => t.def_id_full(c),
+                None => t.def_id(),
+            }
+        }
```

Comment:
nit: maybe use `let t: &dyn GetDefId = match ...` instead, so this isn't monomorphized over and over again.